src/meta/ActiveHostsMan.cpp

```diff
@@ -92,6 +92,40 @@ bool ActiveHostsMan::machineRegisted(kvstore::KVStore* kv, const HostAddr& hostA
   return code == nebula::cpp2::ErrorCode::SUCCEEDED;
 }
 
+ErrorOr<nebula::cpp2::ErrorCode, std::vector<std::pair<HostAddr, cpp2::HostRole>>>
+ActiveHostsMan::getActiveServicesInHost(kvstore::KVStore* kv,
+                                        std::string hostname,
+                                        int32_t expiredTTL) {
+  const auto& prefix = MetaKeyUtils::hostPrefix();
+  std::unique_ptr<kvstore::KVIterator> iter;
+  auto retCode = kv->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+  if (retCode != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    LOG(ERROR) << "Failed to get active services in the host: " << hostname << ", error "
+               << apache::thrift::util::enumNameSafe(retCode);
+    return retCode;
+  }
+
+  int64_t threshold =
+      (expiredTTL == 0 ? FLAGS_heartbeat_interval_secs * FLAGS_expired_time_factor : expiredTTL) *
+      1000;
+  auto now = time::WallClock::fastNowInMilliSec();
+
+  std::vector<std::pair<HostAddr, cpp2::HostRole>> hosts;
+  while (iter->valid()) {
+    auto addr = MetaKeyUtils::parseHostKey(iter->key());
+    HostInfo info = HostInfo::decode(iter->val());
+
+    if (addr.host == hostname) {
+      if (now - info.lastHBTimeInMilliSec_ < threshold) {
```

Comment:
Do we need to return all services regardless of threshold?
IMO, if one storage is crashed, maybe we should not proceed to backup. So as for agent, maybe they should get all services info no matter what state it is. In that way, agent could return error if some services is crashed.

WDYT?

Reply:
You are right.